### PR TITLE
feat: adds postgresql/sqlite destination connector

### PR DESCRIPTION
**Pull request by rvztz**
- Adds a destination connector to upload processed output into a PostgreSQL/Sqlite database instance.
- Users are responsible to provide their instances. This PR includes a couple of configuration examples.
- Defines the scripts required to setup a PostgreSQL instance with the unstructured elements schema.
- Validates postgres/pgvector embedding storage and retrieval

### CODE REVIEW

**Comment by potter-potter** [comment]
Can we make the 2 .sh scripts executable?

**Comment by potter-potter** [comment]
I'm missing 
scripts/sql-test-helpers/create-sql-instance.sh: No such file or directory

scripts/sql-test-helpers/docker-compose.yml: no such file or directory

maybe they haven't been checked into this branch?

File: examples/ingest/sql/ingest.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env bash
+
+# Uploads the structured output of the files within the given S3 path.
+
+# Structured outputs are stored in azure-ingest-output/
```
Comment (potter-potter) [inline examples/ingest/sql/ingest.sh:5]: fix comment

File: examples/ingest/sql/ingest.sh [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+PYTHONPATH=. ./unstructured/ingest/main.py \
+        s3 \
+         --remote-url s3://utic-dev-tech-fixtures/small-pdf-set/ \
+         --anonymous \
+         --output-dir s3-small-batch-output-to-sql \
+         --num-processes 2 \
+         --verbose \
+        --strategy fast \
+        sql \
+        --db_name postgres \
```
Comment (potter-potter) [inline examples/ingest/sql/ingest.sh:19]: I think its db-name

File: test_unstructured_ingest/dest/sql.sh [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+trap cleanup EXIT
+
+# Create sql instance and create `elements` class
+echo "Creating SQL DB instance"
+# shellcheck source=/dev/null
+scripts/sql-test-helpers/create-sql-instance.sh
+wait
+
+PYTHONPATH=. ./unstructured/ingest/main.py \
+  s3 \
```
Comment (potter-potter) [inline test_unstructured_ingest/dest/sql.sh:39]: Looks like we are using `local` instead of s3 now for destination tests to speed things up. Check out the latest 
https://github.com/Unstructured-IO/unstructured/blob/c471ea3cc7c167f1328256ef1719f6cca4849cad/test_unstructured_ingest/dest/delta-table.sh#L31

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,13 @@
+## 0.10.31-dev0
+
+### Enhancements
+
+### Features
+
+**Weaviate destination connector** PostgreSQL connector added to ingest CLI.  Users may now use `unstructured-ingest` to write partitioned data from over 20 data sources (so far) to a PostgreSQL database.
```
Comment (ryannikolaidis) [inline CHANGELOG.md:7]: Weaviate a typo here?

**Comment by ryannikolaidis** [comment]
can you update pr title to conventional commit message?

**Comment by ryannikolaidis** [comment]
can you add comprehensive pr description?

**Comment by ryannikolaidis** [comment]
update sphinx docs?

**Comment by ryannikolaidis** [comment]
appreciate that you're thinking bigger picture about sql general abstraction, but probably need to comb through and clarify the sql vs postgres-specific portions throughout. for instance, currently test-ingest-sql-output.py specifically tests postgres..so let's bump filename to make that clear.

**Comment by potter-potter** [comment]
Can you check that the ./test_unstructured_ingest/dest/sql.sh runs in CI?
I saw this in the last successfull CI tests
--------- SKIPPING SCRIPT sql.sh ---------

https://github.com/Unstructured-IO/unstructured/actions/runs/6848685710/job/18619418567

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,13 @@
+## 0.10.31-dev0
+
+### Enhancements
+
+### Features
+
+**PostgreSQL destination connector** PostgreSQL connector added to ingest CLI.  Users may now use `unstructured-ingest` to write partitioned data from over 20 data sources (so far) to a PostgreSQL database.
```
Comment (potter-potter) [inline CHANGELOG.md:7]: Are there types of Unstructured data sources that can't be written to postgres?

File: unstructured/ingest/connector/sql/connector.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+                data["metadata"]["data_source"]["id"] = data_source_id
+                data["data_source_id"] = data_source_id
+
+            if data.get("metadata", {}).get("coordinates"):
+                coordinates_id = str(uuid.uuid4())
+                data["metadata"]["coordinates"]["id"] = coordinates_id
+                data["coordinates_id"] = coordinates_id
+
+        # Dict as string formatting
+        if record_locator := data.get("metadata", {}).get("data_source", {}).get("record_locator"):
```
Comment (potter-potter) [inline unstructured/ingest/connector/sql/connector.py:103]: snazzy usage of the walrus :=

File: unstructured/ingest/connector/sql/connector.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+    # @DestinationConnectionError.wrap
+    def write_dict(self, *args, json_list: t.List[t.Dict[str, t.Any]], **kwargs) -> None:
+        from sqlalchemy import insert
+
+        logger.info(
+            f"writing {len(json_list)} objects to database url " f"{self.connector_config.db_url} "
+        )
+
+        elements = make_elements_table(
+            self.write_config.table_name_mapping.get(ELEMENTS_TABLE_NAME, ELEMENTS_TABLE_NAME)
```
Comment (potter-potter) [inline unstructured/ingest/connector/sql/connector.py:159]: Is there a specific reason were using .get like this? (if it doesn't find ELEMENTS_TABLE_NAME, it use ELEMENTS_TABLE_NAME). Probably be better to just use ["ELEMENTS_TABLE_NAME"] and  hard fail if there is no ELEMENTS_TABLE_NAME.

**Comment by potter-potter** [comment]
This SQL destination connector is really cool. Works great. 

Will be nice when it's validated on more than just PostgreSQL. SQLite should be easier since we don't have to do a docker container.

**Comment by cragwolfe** [comment]
> This SQL destination connector is really cool. Works great.
> 
> Will be nice when it's validated on more than just PostgreSQL. SQLite should be easier since we don't have to do a docker container.

Sweet! I also think sqlite opens up a lot of possibilities. By the way, SQLite plus sqlite-vss *is* a vectordb! https://github.com/asg017/sqlite-vss :)

**Comment by potter-potter** [comment]
> > This SQL destination connector is really cool. Works great.
> > Will be nice when it's validated on more than just PostgreSQL. SQLite should be easier since we don't have to do a docker container.
> 
> Sweet! I also think sqlite opens up a lot of possibilities. By the way, SQLite plus sqlite-vss _is_ a vectordb! https://github.com/asg017/sqlite-vss :)

Awesome! SQLite is the new SQLite:). ChromaDB is vectordb on top of SQLite (though I think you can substitute another database in production). And its super simple to use. I'm building the destination connector for it.

File: docs/source/ingest/destination_connectors/sql.rst [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+.. tabs::
+
+   .. tab:: Shell - with parameters
+    
+    You can use the connectors with the same parameters used in `sqlalchemy.engine.URL.create <https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.engine.URL.create>`_
+
+      .. code:: shell
+
+        unstructured-ingest \
+            s3 \
```
Comment (rbiseck3) [inline docs/source/ingest/destination_connectors/sql.rst:27]: Can we update this to use the local source connector to match the destination ingest tests?

File: docs/source/ingest/destination_connectors/sql.rst [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+            --strategy fast \
+            sql \
+            --drivername postgresql \
+            --username postgres \
+            --password test \
+            --host localhost \
+            --port 5432 \
+            --database elements
+
+   .. tab:: Shell - with database-uurl
```
Comment (rbiseck3) [inline docs/source/ingest/destination_connectors/sql.rst:42]:
```suggestion
   .. tab:: Shell - with database-url
```

File: examples/ingest/sql/ingest.sh [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+# Uploads the structured output of the files within the given S3 path.
+
+# Structured outputs are stored in a PostgreSQL instance/
+
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+cd "$SCRIPT_DIR"/../../.. || exit 1
+
+PYTHONPATH=. ./unstructured/ingest/main.py \
+        s3 \
```
Comment (rbiseck3) [inline examples/ingest/sql/ingest.sh:11]: Same comment as above, using local connector for the source.

File: requirements/ingest/sql.in [hunk 11]
```diff
@@ -0,0 +1,4 @@
+-c ../constraints.in
+-c ../base.txt
+SQLAlchemy
+psycopg2-binary
```
Comment (rbiseck3) [inline requirements/ingest/sql.in:4]: Is the `psycopg2-binary` dependency specific to postgres or all sql connections?

File: scripts/sql-test-helpers/docker-compose-postgresql.yaml [hunk 12]
```diff
@@ -0,0 +1,12 @@
+services:
+  postgres:
+    image: postgres:14-alpine
```
Comment (rbiseck3) [inline scripts/sql-test-helpers/docker-compose-postgresql.yaml:3]: Would it be possible to leverage an image that has the [pgvector](https://github.com/pgvector/pgvector) extension and store the embeddings in a vector column. Ideally leverage a vector-specific query against it in the ingest destination test.

File: scripts/sql-test-helpers/test-ingest-sql-output.py [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    engine = create_engine(db_url)
+    with engine.connect() as conn:
+        count = conn.execute(text("select count(*) from elements;")).scalar()
+        try:
+            assert count == N_ELEMENTS
+        except AssertionError:
+            print(f"{database_name} dest check failed: got {count}, expected {N_ELEMENTS}")
+            raise
+
+        print(f"{database_name} dest check success")
```
Comment (rbiseck3) [inline scripts/sql-test-helpers/test-ingest-sql-output.py:38]: If possible, adding a test using a vector-specific query against the embedding content.

File: scripts/sql-test-helpers/create-sql-instance.sh [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+set -e
+
+SCRIPT_DIR=$(dirname "$(realpath "$0")")
+DATABASE_NAME=$1
+
+# Create the SQL instance
+if [[ "$DATABASE_NAME" != "sqlite" ]]; then
+    docker compose version
+    docker compose -f "$SCRIPT_DIR"/docker-compose-"$DATABASE_NAME".yaml up --wait
+    docker compose -f "$SCRIPT_DIR"/docker-compose-"$DATABASE_NAME".yaml ps
```
Comment (rbiseck3) [inline scripts/sql-test-helpers/create-sql-instance.sh:12]: How does the elements table being written to get provisioned with the appropriate columns?

File: unstructured/ingest/cli/cmds/sql.py [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+        options = [
+            click.Option(
+                ["--drivername"],
+                default="postgresql",
+                type=click.Choice(SQL_DRIVERS),
+                help="Name of the database backend",
+            ),
+            click.Option(
+                ["--username"],
+                default="postgres",
```
Comment (rbiseck3) [inline unstructured/ingest/cli/cmds/sql.py:34]: Let's avoid setting any defaults specific to one of the drivers.

File: unstructured/ingest/connector/sql/connector.py [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+        self.engine = create_engine(self.connector_config.db_url)
+        if self.write_config.table_name_mapping is None:
+            self.write_config.table_name_mapping = {
+                ELEMENTS_TABLE_NAME: ELEMENTS_TABLE_NAME,
+                METADATA_TABLE_NAME: METADATA_TABLE_NAME,
+                DATA_SOURCE_TABLE_NAME: DATA_SOURCE_TABLE_NAME,
+                COORDINATES_TABLE_NAME: COORDINATES_TABLE_NAME,
+            }
+
+    def check_connection(self):
```
Comment (rbiseck3) [inline unstructured/ingest/connector/sql/connector.py:78]: can this be populated with a valid check?

**Comment by rbiseck3** [comment]
General comments about table creation/management:
* The sqlalchemy code to create tables for the content to be written to is in the core ingest code. We should not include anything in there responsible for provisioning a DB on behalf of a user. We might still need this for integration testing but then that should exist in the test directory.
* We can/should still include a sample schema a user can use to provision their own database. You can provision a table via sqlalchemy and then dump it in a format a user could use with postgres regardless of having access to python. 
* Regarding the schema, in my opinion, we should leverage the json flattening code used by the delta table connector (and any other table-based dbs) when writing the content to avoid the complexities of cross-table joins. This means you only need to provision a single table and write to it.

File: docs/source/ingest/destination_connectors/sql.rst (hunk 8 above)
Comment (potter-potter) [reply docs/source/ingest/destination_connectors/sql.rst:27]: Doesn't seem like this comment has been fulfilled.

File: unstructured/ingest/connector/sql/schema.py [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+                column_name = (
+                    c
+                    if (self.table_column_mapping is None)
+                    else self.table_column_mapping.get(c, c)
+                )
+                columns.append(column_name)
+                values.append(data[c])
+
+        query = (
+            f"INSERT INTO {table_name} ({','.join(columns)}) "
```
[the remark this answers is not in the document]
Comment (potter-potter) [reply unstructured/ingest/connector/sql/schema.py:77]: Are you able to use query parameters? It's how I've usually done it for raw sql. 
https://realpython.com/prevent-python-sql-injection/#using-query-parameters-in-sql

File: test_unstructured_ingest/test-ingest-dest.sh [hunk 18]
```diff
@@ -24,6 +24,8 @@ all_tests=(
   'mongodb.sh'
   'pinecone.sh'  
   's3.sh'
+  'postgresql.sh'
+  'sqlite.sh'
```
Comment (potter-potter) [inline test_unstructured_ingest/test-ingest-dest.sh:28]: shouldn't mysql.sh be here?

File: requirements/ingest/sql.in (hunk 11 above)
Comment (rbiseck3) [reply requirements/ingest/sql.in:3]: then we might want to have this in a `postgres.in` file which extends a `sql.in` file if there are dependencies common across them.

**Comment by potter-potter** [comment]
@rvztz Let's ignore all the mysql comments. We can add those at a different point in time. Let's just try to button up the postgres and sqlite :)

**Comment by potter-potter** [comment]
I'm standardizing on `postgres` instead of `postgresql` where practical.

File: test_unstructured_ingest/python/test-ingest-sql-output.py [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+    cursor = conn.cursor()
+    cursor.execute(query)
+    count = cursor.fetchone()[0]
+
+    if database_name == "pgvector":
+        query = "SELECT AVG(embeddings) FROM elements;"
+        cursor = conn.cursor()
+        cursor.execute(query)
+        res = cursor.fetchone()
+        print(f"Result of {query} against pgvector with embeddings")
```
Comment (rbiseck3) [inline test_unstructured_ingest/python/test-ingest-sql-output.py:47]: Are we not asserting on the results from pgvector?

File: test_unstructured_ingest/unit/test_paths.py [hunk 20]
```diff
@@ -127,3 +133,91 @@ def test_fsspec_folder_fails():
 
     assert output_filename == Path("/fake_file2.txt.json")
     assert download_filename == Path("/fake_file2.txt")
+
+
+def test_post_init_invalid_protocol():
```
Comment (rbiseck3) [inline test_unstructured_ingest/unit/test_paths.py:138]: Are these changes relevant to this PR directly? Or just something added along the way?

File: unstructured/ingest/connector/sql/connector.py [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+
+from .schema import (
+    ELEMENTS_TABLE_NAME,
+    DatabaseSchema,
+)
+
+
+@dataclass
+class SqlAccessConfig(AccessConfig):
+    password: t.Optional[str] = enhanced_field(sensitive=True)
```
Comment (rbiseck3) [inline unstructured/ingest/connector/sql/connector.py:25]: I would add username in here as well, but not mark is as sensitive as it's still a value needed for accessing the db.

File: unstructured/ingest/connector/sql/connector.py [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+
+    @property
+    def connection(self):
+        if self.db_type == "postgresql":
+            return self._make_psycopg_connection
+        elif self.db_type == "sqlite":
+            return self._make_sqlite_connection
+        raise ValueError(f"Unsupported database {self.db_type} connection.")
+
+    def _make_sqlite_connection(self):
```
Comment (rbiseck3) [inline unstructured/ingest/connector/sql/connector.py:52]: Do we need the `required_dependency` decorator here for the `sqlite3` dep?

File: unstructured/ingest/connector/sql/connector.py [hunk 23]
```diff
@@ -0,0 +1,216 @@
...
+            self.write_config.mode == "error" or self.write_config.mode == "ignore"
+        ) and schema_exists:
+            raise ValueError(
+                f"There's already an elements schema ({ELEMENTS_TABLE_NAME}) "
+                f"at {self.connector_config.db_type}"
+            )
+        if self.write_config.mode == "overwrite" and schema_exists:
+            schema_helper.clear_schema()
+
+    # @DestinationConnectionError.wrap
```
Comment (rbiseck3) [inline unstructured/ingest/connector/sql/connector.py:180]: Is still supposed to be commented out?

File: unstructured/ingest/connector/sql/connector.py [hunk 24]
```diff
@@ -0,0 +1,216 @@
...
+            self._resolve_mode(schema_helper)
+            for e in json_list:
+                elem = self.conform_dict(e)
+
+                schema_helper.insert(
+                    ELEMENTS_TABLE_NAME,
+                    elem,
+                )
+
+            conn.commit()
```
Comment (rbiseck3) [inline unstructured/ingest/connector/sql/connector.py:202]: Should there be a limit on how many insert statements we add in for each `commit`?  Not sure if the sdk batches them up for us or not.

File: unstructured/ingest/connector/sql/connector.py [hunk 25]
```diff
@@ -0,0 +1,216 @@
...
+            schema_helper.cursor.close()
+        conn.close()
```
Comment (rbiseck3) [inline unstructured/ingest/connector/sql/connector.py:204]: Does the connection already get closed when the `with ... as ...` block exits?

File: unstructured/ingest/connector/sql/schema.py [hunk 26]
```diff
@@ -0,0 +1,90 @@
+import json
```
Comment (rbiseck3) [inline unstructured/ingest/connector/sql/schema.py:1]: Where is this file used? I'd like to avoid having any code around creating a schema in the core ingest directory. This can live in either the test or the docs directory perhaps.

File: docs/source/ingest/destination_connectors/sql.rst [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+NOTE: Keep in mind that you will need to have all the appropriate extras and dependencies for the file types of the documents contained in your data storage platform if you're running this locally. You can find more information about this in the `installation guide <https://unstructured-io.github.io/unstructured/installing.html>`_.
+
+Sample Index Schema
+-------------------
+
+To make sure the schema of the index matches the data being written to it, a sample schema json can be used.
+
+PostgreSQL:
+
+.. literalinclude:: ./data/postgres-schema.sql
```
Comment (rbiseck3) [inline docs/source/ingest/destination_connectors/sql.rst:43]: This might look nice in dedicated tabs? Similar to how the python and shell code are in dedicated tabs.

File: test_unstructured_ingest/unit/test_sql_conform_dict.py [hunk 28]
```diff
@@ -0,0 +1,125 @@
+import datetime
```
Comment (rbiseck3) [inline test_unstructured_ingest/unit/test_sql_conform_dict.py:1]: Appreciate all the unit tests.

File: unstructured/ingest/cli/cmds/sql.py [hunk 29]
```diff
@@ -0,0 +1,66 @@
...
+SQL_DRIVERS = {"postgresql", "sqlite"}
+
+
+@dataclass
+class SqlCliConfig(SimpleSqlConfig, CliConfig):
+    @staticmethod
+    def get_cli_options() -> t.List[click.Option]:
+        options = [
+            click.Option(
+                ["--db_type"],
```
Comment (rbiseck3) [inline unstructured/ingest/cli/cmds/sql.py:19]:
```suggestion
                ["--db-type"],
```

**Review by rbiseck3 — APPROVED** [review]
A couple small comments but otherwise LGTM
